### ISSUE TITLE
Polygon and Point extrusion

### ISSUE THREAD

**Issue by d-bose** [openlayers/ol-cesium#515]
Polygon and Point extrusion

Questions on this topic has been asked few times in the forum. But I could not get my polygon to get extruded as you would find in this cesium link
https://cesiumjs.org/Cesium/Apps/Sandcastle/gallery/Geometry%20and%20Appearances.html
I want to raise up/extrude my polygons to different heights
I was expecting an example in this link (vectors example)
http://openlayers.org/ol-cesium/examples/vectors.html
but I could not find. This link has examples of polygons that are floating above ground but they are not connected to the ground. I played around with the code locally but could not manage to get any polygon to extrude out from the ground. 
I looked at the following links and tried some code but could not manage to extrude my polygon
https://github.com/openlayers/ol-cesium/issues/167
https://github.com/openlayers/ol-cesium/issues/360
https://github.com/gberaudo/talks/tree/master/2015-foss4g-seoul/examples/extrude
I tried with the example in the following link but that gave be extruded polygon only in the 3D view. There was no polygon in the 2D view
https://github.com/AnalyticalGraphicsInc/cesium/blob/8a9edfda67d859a6ce5aa4d20cc5c5cf238ffd62/Apps/Sandcastle/gallery/Geometry%20and%20Appearances.html

I am looking for an example that will show an extruded polygon in 3D but when seen it 2D will show up as a flat polygon. Just like in the vectors example link above where all the objects show up in both 2D & 3D view. I also need my point data, like circles to be extruded. Thanks for any help.

**Comment by gberaudo** [comment]
Hi @d-bose, yes there is no support for extrusion in OL-Cesium yet. There are 2 reasons for it:
- I never really had a use of it (only for the building POC);
- there is no concept of extrusion in OpenLayers / many ways to encode extrusion information

How to implement it?
We could:
- add an `olcs_get_extrusion` function in each property / layer that would return the base_height of the feature and the extrusion_height;
- check for the presence of this property in the feature converter and pass the information to Cesium.

I assume animating extrusion is unneeded. Changing the extrusion would be possible at the cost of recreating the 3D feature (through a change event).

This would work great for polygons. You are also writting about extruding points; what do you mean? you want a line under the point? It would be possible by adding it to the primitive returned by `olPointGeometryToCesium` but we need a style; it could be also returned by the extrusion function.

Do I understand correctly your need? Are you willing to work on this? If yes I can give you guidance.

**Comment by d-bose** [comment]
Hi @gberaudo 
Thanks for your reply. Yes I am willing to work on this. Please provide guidance. 
This is one of the main usecases because of which we are considering 3D visualization on Map. We have been into 2d map based visualization for more than a decade. We wanted to convey more information on the map & we have seen that bringing height to a point can convey one extra information about that plotting apart from color/shape/size, etc. Same goes for polygons. Hence, extrusion is very important for us. 
Yes, animating extrusion is not needed.
We depict point on our 2D map with circle, square, triangle, etc. On 3D we wanted to show them as towers, like, circle in 2D would become a cylinder in 3D, a square will become a box, and so on.. you can find such examples in cesium (https://cesiumjs.org/Cesium/Apps/Sandcastle/?src=Box.html&label=Geometries)

**Comment by gberaudo** [comment]
@d-bose, do you mean you are using instanced of https://openlayers.org/en/latest/apidoc/ol.style.RegularShape.html in 2d?

**Comment by d-bose** [comment]
No, I have not started using anything. I was evaluating ol-cesium to see whether it meets our needs by referring to your vector example (http://openlayers.org/ol-cesium/examples/vectors.html)

**Comment by d-bose** [comment]
@gberaudo 
I have been trying to go by your suggestion. I cloned the code to make some changes in FeatureConverter. However, I am facing problems in building. I am using a windows 10 machine. Building with windows command prompt gave me errors like "**_mkdir -p dist/ The syntax of the command is incorrect._**" After that I tried building with **_cygdrive_** but getting error in that too - **_"ERR! Could not find base.js"_** & I think it is coming from "_ol-cesium\node_modules\@camptocamp\closure-util\lib\manager.js_". Can you please help me with that? 

Alternatively, I was trying to get hold of the **_olcesium-debug.js_** But it seem that the file is not there. As per your website it is writen under - 
**_Running the unminified version of Cesium -. Passing the parameter ?mode=dev to an example will load the debug version of Cesium instead of the minified one._** 
However, the links seem to be broken. - (http://openlayers.org/ol-cesium/olcesium-debug.js). Can you make that available so that the examples can be run in debug mode?

Also, can you make the the "dist" folder available in git?

**Comment by gberaudo** [comment]
Sorry for the late reply, see #519.

**Comment by alvaroind** [comment]
Hi all!, in which state is the conversion of polygons to polygons / boxes in 3D? Is it within the last release?

**Comment by gberaudo** [comment]
Hi @d-bose, the build system as completely changed since your last comment. It is now based on webpack and should now work also on windows.

@alvaroind, there is no conversion per se, but you can specify what (existing) 3d model you want to use instead of the 2d shape.

**Comment by gberaudo** [comment]
See https://github.com/openlayers/ol-cesium/blob/246e806d5966e90e8ba8f376a9bdc9b2aa8aeeb1/src/olcs/FeatureConverter.js#L796-L808.

**Comment by alvaroind** [comment]
Thanks for your reply @gberaudo , my question is if I could represent polygon from 2d (in the image, red polygon) to, for example, polygon or a box in 3D (in the image, blue box, this is created only in Cesium, it is not appear in 2D map with openlayers). How to represent a polygon from 2D to 3D?

Thanks in advance.

![image](https://user-images.githubusercontent.com/37860823/46393857-414b5080-c6e7-11e8-81e6-cd4e395d2027.png)

**Comment by gberaudo** [comment]
Out of the box your 2d polygons should appear on the globe. If you use terrain, you must set the `altitudeMode` to `clampToGround`.

For extrusion there is currently no support "out of the box" in OL-Cesium. What you can do is overriding methods in the FeatureConverter class:
- create a class that extends the FeatureConverter;
- override the methods you want behave differently;
- create the synchronizers and instantiate olcs.OLCesium with them.

**Comment by fredj** [comment]
see #1038
